GPflowOpt/design.py

```diff
@@ -67,3 +74,147 @@ def __init__(self, domain):
 
     def generate(self):
         return np.empty((0, self.domain.size))
+
+
+class LatinHyperCube(Design):
+    """
+    Latin hypercube with optimized maximin distance. Created with the Translational Propagation algorithm to obtain 
+    some speed and avoid lengthy generation procedures. For dimensions smaller or equal to 6, this algorithm finds 
+    the optimal LHD (or gets very close) with overwhelming probability. 
+    Beyond 6D, this property fades, although the resulting designs are still acceptable. Somewhere beyond 15D this 
+    algorithm tends to slow down a lot. Key reference is
+    
+    ::
+       @article{Viana:2010,
+            title={An algorithm for fast optimal Latin hypercube design of experiments},
+            author={Viana, Felipe AC and Venter, Gerhard and Balabanov, Vladimir},
+            journal={International Journal for Numerical Methods in Engineering},
+            volume={82},
+            number={2},
+            pages={135--156},
+            year={2010},
+            publisher={John Wiley & Sons, Ltd.}
+       }
+
+    """
+
+    def __init__(self, size, domain, max_seed_size=None):
+        super(LatinHyperCube, self).__init__(size, domain)
+        self._max_seed_size = max_seed_size or domain.size
+
+    def generate(self):
+        # Generate several TPLHDs with growing seed, select the one with the best intersite distance
+        candidates = []
+        scores = []
+
+        for i in np.arange(1, min(self.size, self._max_seed_size) + 1):
+            if i < 3:
+                # Hardcoded seeds for 1 or two points.
+                seed = np.arange(1, i + 1)[:, None] * np.ones((1, self.domain.size))
+            else:
+                # Generate larger seeds recursively by creating small LHD's
+                seed = self._tplhs_design(i, np.ones((1, self.domain.size)))
+
+            # Create all designs and compute score
+            X = self._tplhs_design(self.size, seed)
+            candidates.append(X)
+            scores.append(np.min(pdist(candidates[-1])))
+
+        # Transform best design (highest score) to specified domain
+        cube = np.sum([ContinuousParameter('x{0}'.format(i), 1, self.size) for i in np.arange(self.domain.size)])
```

Comment:
Hmm, I think we should let the scaling be done by the user if possible?

I was already thinking of making cube a property of the base class, and hence let the base class do this scaling. But in the end just let the user do the scaling: the domain classes are easy to use specifically for that reason? It will also make the whole discrete/continuous parameters easier.